File: ollama/codigo-final/binary-trees/llama3.1-8b/cot/temp_0/binary-trees.cpp
#include <iostream>
#include <cstdlib>

// Define a tree node structure with left and right child pointers
struct TreeNode {
    int value;
    TreeNode* left;
    TreeNode* right;

    TreeNode(int val) : value(val), left(nullptr), right(nullptr) {}
};

// Function to allocate a stretch tree of depth D
TreeNode* allocateStretchTree(int D) {
    if (D == 0) return nullptr;
    TreeNode* node = new TreeNode(D);
    node->left = allocateStretchTree(D - 1);
    node->right = allocateStretchTree(D - 1);
    return node;
}

// Function to check the existence of a tree
bool checkTree(TreeNode* root, int D) {
    if (root == nullptr || D < 0) return false;
    bool leftExist = checkTree(root->left, D - 1);
    bool rightExist = checkTree(root->right, D - 1);
    return leftExist && rightExist;
}

// Function to deallocate a tree
void deallocateTree(TreeNode* root) {
    if (root == nullptr) return;
    deallocateTree(root->left);
    deallocateTree(root->right);
    delete root;
}

int main(int argc, char** argv) {
    int N = std::stoi(argv[1]);
    int min_depth = 4;
    int max_depth = std::max(6, N);

    // Allocate a stretch tree of depth D
    for (int D = min_depth; D <= max_depth; D += 2) {
        TreeNode* stretchTree = allocateStretchTree(D);
        bool checkResult = checkTree(stretchTree, D);
        deallocateTree(stretchTree);
        std::cout << "stretch tree of depth " << D << "\t check: " << (checkResult ? "C" : "F") << std::endl;
    }

    // Allocate a long-lived tree
    TreeNode* longLivedTree = allocateStretchTree(max_depth);

    // Allocate, walk, and deallocate many bottom-up trees
    for (int D = min_depth; D <= max_depth; D += 2) {
        int iterations = 1 << (max_depth - D + min_depth);
        for (int i = 0; i < iterations; ++i) {
            TreeNode* tree = allocateStretchTree(D);
            deallocateTree(tree);
        }
    }

    // Ensure the long-lived tree still exists
    bool checkResult = checkTree(longLivedTree, max_depth);
    std::cout << "long lived tree of depth " << max_depth << "\t check: " << (checkResult ? "C" : "F") << std::endl;

    return 0;
}

File: ollama/codigo-final/binary-trees/llama3.1-8b/cot/temp_0/binary-trees.py
import os

class Node:
    def __init__(self):
        self.left = None
        self.right = None

def allocate_stretch_tree(depth, max_depth):
    if depth > max_depth:
        return None
    node = Node()
    node.left = allocate_stretch_tree(depth + 1, max_depth)
    node.right = allocate_stretch_tree(depth + 1, max_depth)
    return node

def check_long_lived_tree(tree):
    return tree is not None

def walk_bottom_up_tree(node, depth):
    if node is None:
        return
    print(f"Depth: {depth}")
    walk_bottom_up_tree(node.left, depth + 1)
    walk_bottom_up_tree(node.right, depth + 1)

def allocate_long_lived_tree():
    tree = Node()
    return tree

def main(min_depth=4, max_depth=max(6, N), stretch_depth=max_depth + 1):
    iterations = 1 << (max_depth - min_depth)
    
    # Allocate a long-lived tree
    long_lived_tree = allocate_long_lived_tree()

    for depth in range(min_depth, max_depth + 1, 2):
        print(f"Allocating stretch tree of depth {depth}")
        stretch_tree = allocate_stretch_tree(0, stretch_depth)
        
        # Check that the long-lived tree still exists
        assert check_long_lived_tree(long_lived_tree)

        # Walk and deallocate bottom-up trees
        for i in range(iterations):
            print(f"Allocating bottom-up tree of depth {depth}")
            node = Node()
            walk_bottom_up_tree(node, 0)
            
            print(f"Deallocating bottom-up tree of depth {depth}")

    # Check that the long-lived tree still exists after all allocations
    assert check_long_lived_tree(long_lived_tree)

if __name__ == "__main__":
    N = int(os.environ.get("N", "10"))  # Get value from environment variable or default to 10
    main(min_depth=4, max_depth=max(6, N), stretch_depth=max_depth + 1)
